### ISSUE TITLE
issue about detector accuracy

### ISSUE THREAD

**Issue by yochiawu** [umyelab/LabGym#196]
Hi,

I would like to ask how to improve the accuracy of the detector. I am using LabGym2.4.5 to detect the behavior of two mice. The video fps is 30 and the framesize is 720*1280. However, there are many circling behavior and overlap in the video so it is difficult to detect(Like the figure below). The data of the best detector I have trained is: 176 training set, framesize:960, iteration: 1000. But it still has two switches in 30s. I tried to change the framesize(from 640 to 1280), iteration(from 1000 to 6000), training set(from 176 to 292), but none of them works. 

Could you please give me some advice? Thanks a lot!

![2-1-tailmarker_processed_960](https://github.com/user-attachments/assets/97673bf7-49cc-491f-aef3-39b8d76a1198)

**Comment by yujiahu415** [comment]
Hi,

First, since you already had an annotated dataset, the simplest attempt to increase accuracy is to try these: 1. if you use any preprocessing steps like auto orient or resize in Roboflow, remove them; 2. when perform augmentation, for images like you showed, don't do "vertical flipping" as the mouse upside down would never happen in real scenario, instead, only do "horizontal flipping". You may also do "brightness" as the contrast between mouse and background is low and "brightness" might increase the detection accuracy of the trained Detector. 3. go 8000-10000 iterations when training the Detector, or until the "total loss" in training drop below 0.1.

If this doesn't work, you may need to refine your annotation. First, if possible, increase the contrast of the video, you can use LabGym preprocessing module to achieve this. Resize your videos to 960. 1280 may not be necessary. Keep frame size of all videos 960, and use 960 as the inferencing frame size when training the Detector, go 8000-10000 iterations. Generate images at 960 frame size, and select more colliding/occlusion scenarios. I don't know what are the portions of your current 176-292 images, and don't know whether this number is before or after augmentation. For complex scenarios like low contrast and frequent occlusions, we generally recommend 200~300 annotated images before augmentation. Importantly, make the ratio of colliding/occlusion scenarios 97~99% and the well-separated scenarios 3~1% of the total annotated images.

If there's still ID switching after you do all the above, you may need to either identify the frames which ID switchings happen to include them in the training image set, or record the videos with higher resolution and contrast, or use overhead cameras so that the scenarios that one mouse is completely occluded by another happen less frequently.

**Comment by yochiawu** [comment]
Thanks a lot for your advice!

Recently, I am using video recorded by a overhead camera to train the detector. The frame size is 960, training set is 203 annotated images with 6 well-separated before augmentation, the iteration is 10000 (it seems like I can't set it higher). But the total loss is always above 0.1 (around 0.12). Does it matter?

The detector works well for most of the time but there is still one ID switch which I really don't like. And actually the training set contained some of these frames. The scenario can be seen as below. It is one mouse on top of the other. So the mouse on the bottom is devided into two parts. I am wondering if there is anything wrong with my annotation.

MY ANNOTATION
![image](https://github.com/user-attachments/assets/96db0255-ebb0-4fc2-ae35-676c86940002)
![image](https://github.com/user-attachments/assets/b1364efa-430e-48e9-bd60-2e9d05a9f473)


ID SWITICH SCENARIO (frames from behavior example)
![image](https://github.com/user-attachments/assets/af18f693-cbae-4527-b70e-7be1b2252e8b)
![image](https://github.com/user-attachments/assets/c2663e2b-dd39-4f3a-8e7b-8d35f9608128)
![image](https://github.com/user-attachments/assets/9742e42a-ba80-453b-aea4-5370f60e2b57)
![image](https://github.com/user-attachments/assets/104f6799-40b8-4345-b9a0-d69564b71a55)
![image](https://github.com/user-attachments/assets/55c80ceb-302f-41f9-8d31-9aa0fdac8485)
![image](https://github.com/user-attachments/assets/ed01e362-152a-448e-841d-d0bd6d3bddd6)

Thanks again!

**Comment by yochiawu** [comment]
Hi,

Yes, I annotate the separate parts as one mouse. The augmentations contain: Flip: Horizontal, Vertical; Brightness: Between -15% and +15%; Exposure: Between -10% and +10%.

Thanks for explaining to me that. If I annotate just one part of the bottom mouse, should I always annotate head or rear or always annotate the larger part. Because the top mouse would move on the bottom one, sometimes the head is covered and sometimes the rear is covered. I don't know if it will confuse the Detector if I annotate the head sometimes and the rear sometimes.

**Comment by yujiahu415** [comment]
Here's a trick to increase the amount and diversity of your training dataset in Roboflow:
For Roboflow free account, it expand the number of augmented images to around 2 folds of the original images. Therefore, if you choose multiple augmentation methods (>2) at once, not all images will be applied to a method. To maximize the chance of each augmentation method being applied to each image, you can select one method at a time, and export the dataset, and then upload it to create a new project, then apply another method.. and so on. By this way, you will have a dataset of around at least 8X more images (around 1600 images) than the original dataset with those three methods, which could potentially benefit your training and increase the detection accuracy of your Detectors.

As for the annotation, you can always annotate the larger part. When a mouse is occluded, its body isn't intact anyway, so the goal for the Detector in such scenarios is not to be precise, but to be sensitive so that it still detects something that can let it maintain the correct ID.

**Comment by yochiawu** [comment]
Thanks a lot! I tried what you suggested and the accuracy of detector has improved!

But now I am stuck with the accuracy of the categorizer. I am wondering could you help me with that?

I am training an 'interactive advanced' mode. But actually, I am focusing in 2 types of interactive behavior and 2 types of non-interactive behavior. So the first question is, should I train 2 different modes (one interactive advanced and one non-interactive) to categorize them?

When I trained the interactive advanced mode, the behavior example contains 15 frames. Should I reduce this to something like 10 frames to improve accuracy, or does it not matter.

Just focusing on the interactive behavior, sniffing can't be categorized well.  Sniffing is one mouse sniff to another and it contains physical contact. There is 2 major mistakes. The first one is that even if the animal is just walking around without any contact, it would be seen as sniffing. I thought it might relate with the social distance, but If I try to set it smaller as 1 when I generate the behavior example, in many sniffing scenarios there is just one animal which I don't want it to be. So I can only set it as 2 which contains some scenarios that the 2 animals are totally separated. The second one is sometimes the animal be sniffed would be labled as sniffing, but I only sort the animal performing sniffing.

And I am wondering should I sort the behavior into many different sub-groups? For example, now I just sort it as sniffing. But actually there are different types of sniffing. Would that help if I sort it as staying&sniffing, moving&sniffing, staying&besniffed, moving&besniffed?

I know this is a lot of questions. Thanks for your time and any advice would be appreciated!

**Comment by yujiahu415** [comment]
Hi,

First of all, you only need to train one Categorizer with "interactive advanced" mode, because it can categorizer both interactive behaviors and non-interactive behaviors. Social distance=2 would be appropriate. Setting the duration of one behavior example as 15 frames with fps of 30 would also be good for most mouse behaviors.

Generally speaking, to categorize a behavior well, the behavior examples of that category should be very distinct from other categories. That is, when you watch the behavior examples, you can easily see the differences between that category and others. Besides, the behavior examples in the same category should share some similar features. If the behavior examples in the same category look very different, the Categorizer would not learn well. So if you think the sniffing has multiple types that are very different, it'd be better to divide it into multiple categories.

Specifically, I would imagine that staying&sniffing would look very different from moving&sniffing so I suggest you to have those two distinct categories. By the way, how many pairs of behavior examples for each category did you selected? Are they diverse enough to cover most of the behavior variant, or many of them are just redundant/repeats that look identical? If sniffing sometimes was misclassified with being sniffed, you probably need to increase the amount and diversity of the behavior examples in both categories. And also, how many categories in total did you sort? Categorizing the background behaviors well would also be helpful to increase the accuracy of categorizing the behaviors of your interest.

**Comment by yochiawu** [comment]
Hi,

Thanks for your advice! I sorted 7 behavior, and the most important behavior contains 600 examples. But I think they look similar. I will try to enrich the dataset. Thanks a lot!

**Comment by yujiahu415** [comment]
To train a Categorizer that can generalize well, you can sparsely select behavior examples from different animals across different videos, rather than selecting all examples from just one video. For interactive advanced mode, we recommend 300-500 pairs of behavior examples per category to train a good Categorizer. You may take a look at the sorted behavior examples in [LabGym Zoo](https://github.com/yujiahu415/LabGym/blob/master/LabGym_Zoo.md) and get an idea on how they should look like in each category.

**Comment by yochiawu** [comment]
Thanks for the examples!

But I still have a concern. Using walking as an example, if I set the social distance to 2, there are two scenarios. One is when there is only one animal walking, and the other is when there is one animal walking, but another animal appears in the graph. This also happens to other non-interactive behaviors. Should I sort them separately? Like walking alone and walking near others?
![Video_2160 0-3360 0_processed_mice-6hFC_1_3690_len15_scdt2_itadv](https://github.com/user-attachments/assets/669e243a-145d-4f77-9342-57e979579a5c)
![2-3_2183-3520_processed_mice-6hFC_1_14250_len15_scdt2_itadv](https://github.com/user-attachments/assets/aae6b72b-7908-4672-9c2e-f7a97d4edf47)

**Comment by yujiahu415** [comment]
Hi,

Generally speaking, if you include enough examples of both scenarios, the trained Categorizer can learn to recognize the key pattern of moving (colored contours) and consider the other mouse (the gray contours) as background noise.

If you take a look at [our sorted behavior examples of the two-vole interaction](https://drive.google.com/file/d/15EMsQyS1ZH12UsvrakPufeL8t24MQ7xv/view?pli=1), you would notice we also included both scenarios in 'locomotion' and [the trained Categorizer](https://drive.google.com/file/d/1x_wLN494PEILFOAXcgyW4xhssiU4U1W3/view) is pretty accurate in detection this type of behavior. By the way, the 'investigating' and 'allowing' in that behavior example dataset is pretty similar to the 'sniffing' and 'being sniffed' in your experiment and the trained Categorizer (also top view, 2 animals, 30 fps, 9 behaviors) may generalize on your data. You probably can take a look at the examples and try that Categorizer and get some idea about how to sort the behavior examples and how to train a Categorizer.
